tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,279 @@
+      return res;
+    } catch (e) {
+      console.log(e);
+      return null;
+    }
+
+  },
+
+  /**
+   * Get the number of days the user has been on their period
+   * @return {Promise} Resolves into 0 if user not on period, and an integer of the days they have been on their period otherwise
+   */
+  GETPeriodDay: async function (){
+
+    let periodDays = 0;
+    var date = new Date()
+    let dateSymptoms = await GETsymptomsForDate(date.getDate(), date.getMonth()+1, date.getFullYear());
+    if (dateSymptoms.flow === null || dateSymptoms.flow === FLOW_LEVEL.NONE){
+      return 0;
+    }
+    else {
+      let startDate = await this.GETMostRecentPeriodStartDay();
+      return getDaysDiff(startDate, date) ;
+    }
+
+    return periodDays;
+
+  },
+
+
+  /**
+   * Get most recent period start date
+   * @return {Promise} A promise that resolves into a Date object that is when the most recent period started.
+   */
+  GETMostRecentPeriodStartDay: async function () {
+    var date = new Date()
+
+    let mostRecentPeriodDay = getLastPeriodStart(date);
+    return mostRecentPeriodDay;
+  },
+
+  /**
+   * Get how far the user is into their period as a percentage
+   * @return {Promise} A percentage approximation (meaning in range [0,1]) of how far the user is into their period
+   */
+  GETCycleDonutPercent: async function() {
+    try{
+      let today = new Date();
+      let today_str = getDateString(today);
+      let percent = await AsyncStorage.getItem(Keys.CYCLE_DONUT_PERCENT)
+      percent = percent != null ? JSON.parse(percent) : null;
+
+
+
+      if (percent != null && today_str in percent){
+        return percent[today_str];
+      }
+      else{
+        let mostRecentPeriodStart = await this.GETMostRecentPeriodStartDay();
+        let avgCycleLength = await this.GETAverageCycleLength();
+        if (mostRecentPeriodStart && avgCycleLength){
+          let daysSincePeriodStart = getDaysDiff(mostRecentPeriodStart, today);
+          let cycleDonutPercent = daysSincePeriodStart / avgCycleLength;
+          this.POSTCycleDonutPercent(cycleDonutPercent);
+          return cycleDonutPercent;
+        }
+        else{
+          //TODO: is 0 the best value to default to? idk
+          return 0;
+        }
+      }
+
+    } catch(e){
+      console.log(e);
+
+    }
+  },
+
+  /** Get the start and length of each period in the given year
+   * @param {number} year The year to retrieve history for
+   * @return {Promise} an object that contains intervals of the user's period (start & length) in that year
+   */
+  GETCycleHistoryByYear: async function(year) {
+    let intervals = []
+    let endOfYear = new Date(year,11,31);
+    let isYearsLastPeriod = true;
+
+    let current = endOfYear;
+    try {
+      // Search backwards until date switches to the previous year
+      while(current.getFullYear() === year){
+        let currentSymptoms = await GETsymptomsForDate(current.getDate(), current.getMonth()+1, current.getFullYear());
```

Comment:
I think we should load an entire year's worth of data from asyncStorage once and then traverse it rather than traverse while calling GETSymptomsForDate potentially ~365 times. 

Propose we add a helper method ```GETSymptomsForYear(year: number)``` to helpers.js where we use .map() to turn every day entry in the Calendar into a Symptoms obj. Then traverse through it. Map docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map 
```
 // This is not tested so syntax may not be 100% but the general structure is here
export const GETSymptomsForYear = async (year) => {
  // get year
  const yearData = JSON.parse(await AsyncStorage.getItem(year.toString()));

  // Check since yearData could be null 
  if (yearData) {
     // .map() basically transforms each item in the array. In this case we turn it all to Symptom objs
      yearData.map((month) => {
          month.map((day) => {
              return day ? new Symptoms(day) : new Symptoms();
          })
      })
     return yearData; 
  }
  return null;
}
```

This would be a pretty big upheaval from what you have now, open to discuss further! 

@edwardhuahan you probably have to traverse through calendar as well to display filter views, do you think this helper function would help

Reply:
oh yeah that's def better. I'll make the swap